packages/react-instantsearch-hooks/src/lib/__tests__/useSearchState.test.tsx

```diff
@@ -0,0 +1,174 @@
+  test('returns the ui state', async () => {
+    const wrapper = createInstantSearchTestWrapper();
+    const { result, waitForNextUpdate } = renderHook(() => useSearchState(), {
+      wrapper,
+    });
+
+    // Initial render state from manual `getWidgetRenderState`
+    expect(result.current).toEqual({
+      uiState: {
+        indexName: {},
+      },
+      indexUiState: {},
+      setUiState: expect.any(Function),
+      setIndexUiState: expect.any(Function),
+    });
+
+    const setUiState = result.current.setUiState;
+    const setIndexUiState = result.current.setIndexUiState;
+
+    await waitForNextUpdate();
+
+    // InstantSearch.js state from the `render` lifecycle step
+    expect(result.current).toEqual({
+      uiState: {
+        indexName: {},
+      },
+      indexUiState: {},
+      setUiState,
+      setIndexUiState,
+    });
+  });
+
+  test('returns the ui state with initial state', async () => {
+    const wrapper = createInstantSearchTestWrapper({
+      initialUiState: {
+        indexName: {
+          query: 'iphone',
+        },
+      },
+    });
+    const { result, waitForNextUpdate } = renderHook(() => useSearchState(), {
+      wrapper: ({ children }) =>
+        wrapper({
+          children: (
+            <>
+              <SearchBox />
+              {children}
+            </>
+          ),
+        }),
+    });
+
+    // Initial render state from manual `getWidgetRenderState`
+    expect(result.current).toEqual({
+      uiState: {
+        indexName: { query: 'iphone' },
+      },
+      indexUiState: { query: 'iphone' },
+      setUiState: expect.any(Function),
+      setIndexUiState: expect.any(Function),
+    });
+
+    await waitForNextUpdate();
+
+    // InstantSearch.js state from the `render` lifecycle step
+    expect(result.current).toEqual({
+      uiState: {
+        indexName: { query: 'iphone' },
+      },
+      indexUiState: { query: 'iphone' },
+      setUiState: expect.any(Function),
+      setIndexUiState: expect.any(Function),
+    });
+  });
+
+  test('returns a function to modify the whole state', async () => {
+    function App() {
+      const { uiState, setUiState } = useSearchState();
+
+      return (
+        <>
+          <button
+            type="button"
+            onClick={() => {
+              setUiState({ indexName: { query: 'hey' } });
+            }}
+          >
+            {uiState.indexName.query}
+          </button>
+          <SearchBox />
+        </>
+      );
+    }
+
+    const { container } = render(
+      <InstantSearchHooksTestWrapper>
+        <App />
+      </InstantSearchHooksTestWrapper>
+    );
+
+    await wait(0);
+
+    expect(container.innerHTML).toMatchInlineSnapshot(
+      `"<button type=\\"button\\"></button>"`
+    );
+
+    userEvent.click(container.querySelector('button')!);
+
+    await wait(0);
+
+    expect(container.innerHTML).toMatchInlineSnapshot(
+      `"<button type=\\"button\\">hey</button>hey"`
+    );
+  });
```

Comment:
The test is easier to read this way, there's no escaping and we understand the intent of clicking the button.

```suggestion
  test('returns a function to modify the whole state', async () => {
    function App() {
      const { uiState, setUiState } = useSearchState();

      return (
        <>
          <button
            type="button"
            onClick={() => {
              setUiState({ indexName: { query: 'new query' } });
            }}
          >
            {uiState.indexName.query}
          </button>
          <SearchBox />
        </>
      );
    }

    const { getByRole } = render(
      <InstantSearchHooksTestWrapper>
        <App />
      </InstantSearchHooksTestWrapper>
    );
    const button = getByRole('button');

    await wait(0);

    expect(button).toHaveTextContent('');

    userEvent.click(button);

    await wait(0);

    expect(button).toHaveTextContent('new query');
  });
```